token/services/interop/state/fabric/ssp.go

```diff
@@ -0,0 +1,201 @@
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package fabric
+
+import (
+	"sync"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/fabric"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/fabric/services/weaver"
+	"github.com/hyperledger-labs/fabric-token-sdk/token"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/interop/state/driver"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/logging"
+	"github.com/pkg/errors"
+	"go.uber.org/dig"
+)
+
+const (
+	QueryPublicParamsFunction = "queryPublicParams"
+
+	DefaultFabricNetworkName = ""
+)
+
+var logger = logging.MustGetLogger("token-sdk.state")
+
+type RelayProvider interface {
+	Relay(fns *fabric.NetworkService) *weaver.Relay
+}
+
+type StateServiceProvider struct {
+	drivers       map[StateDriverName]NamedStateDriver
+	fnsProvider   *fabric.NetworkServiceProvider
+	relayProvider RelayProvider
+	tmsProvider   *token.ManagementServiceProvider
+
+	mu             sync.RWMutex
+	queryExecutors map[string]driver.StateQueryExecutor
+	verifiers      map[string]driver.StateVerifier
+}
+
+func NewStateServiceProvider(
+	drivers map[StateDriverName]NamedStateDriver,
+	fnsProvider *fabric.NetworkServiceProvider,
+	weaverProvider RelayProvider,
+	tmsProvider *token.ManagementServiceProvider,
+) *StateServiceProvider {
+	return &StateServiceProvider{
+		drivers:        drivers,
+		fnsProvider:    fnsProvider,
+		relayProvider:  weaverProvider,
+		tmsProvider:    tmsProvider,
+		mu:             sync.RWMutex{},
+		queryExecutors: map[string]driver.StateQueryExecutor{},
+		verifiers:      map[string]driver.StateVerifier{},
+	}
+}
+
+func (f *StateServiceProvider) RegisterDriver(driver NamedStateDriver) {
+	f.drivers[driver.Name] = driver
+}
+
+func (f *StateServiceProvider) QueryExecutor(url string) (driver.StateQueryExecutor, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	qe, ok := f.queryExecutors[url]
+	if ok {
+		return qe, nil
+	}
+
+	// Fetch public parameters, if not fetched already
+	ppRaw, err := f.fetchPublicParameters(url)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed fetching public parameters from [%s]", url)
+	}
+	pp, err := f.tmsProvider.PublicParametersFromBytes(ppRaw)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed unmarshalling public parameters from [%s]", url)
+	}
+
+	driver, ok := f.drivers[StateDriverName(pp.Identifier())]
+	if !ok {
+		return nil, errors.Errorf("invalid public parameters type, got [%s]", pp.Identifier())
+	}
+	qe, err = driver.Driver.NewStateQueryExecutor(url)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed instantiating state query executor from [%s]", url)
+	}
+	v, err := driver.Driver.NewStateVerifier(url)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed instantiating state verifier from [%s]", url)
+	}
+	f.queryExecutors[url] = qe
+	f.verifiers[url] = v
+
+	return qe, nil
+}
+
+func (f *StateServiceProvider) Verifier(url string) (driver.StateVerifier, error) {
```

Comment:
Is the verifier a sub-case of a query executor? Because eventually we call a method on the chaincode.